src/http/routing.rs

```diff
@@ -27,23 +27,24 @@ enum RatelimitingKind {
 macro_rules! routes {
     ($lt:lifetime, {
         $(
-            $name:ident $({ $($field_name:ident: $field_type:ty),* })?,
+            $name:ident $({ $($field_name_route:ident: $field_type:ty | $field_type_owned:ty),* })?,
```

Comment:
Instead of making the macro syntax messier, can you make this use traits instead? I'm pretty sure `ToOwned` should handle this, but if it doesn't you can just write a simple private trait.

Reply:
How? I somehow need to know the type, which the OwnedRoute should use. I do not know, how a trait would be able to do that.
The problem is the "&str" type. Can you please provide an example for how to do it?

Reply:
The owned type is simply `<$field_type as ToOwned>::Owned`, and if that doesn't return the correct types you can duplicate the definition of `ToOwned` and implement it for the types needed.

Reply:
Will do that.

Reply:
The `field_type_owned` is actually needed, I did not find a way to get compiling code without it.

Reason:
The macro would expand into something like this (irrelevant code removed):
```rust
#[derive(Clone, Debug)]
pub enum OwnedRoute {
    ChannelMessageReaction { channel_id: <ChannelId as ToOwned>::Owned, message_id: <MessageId as ToOwned>::Owned, user_id: <UserId as ToOwned>::Owned, reaction: <&'a str as ToOwned>::Owned},
}
```

`<&'a str as ToOwned>::Owned` is not valid, because `'a` does not exist. I could add the lifetime to the enum, but then we would need to carry this useless lifetime though the whole code. This can not be fixed by replacing the `<&'a str as ToOwned>::Owned` with a custom macro, I already tried:
```rust
macro_rules! get_owned_type {
    (&'a str) => {String};
    ($type:ty) => {<$type as ToOwned>::Owned};
}

get_owned_type!(&'a str);
```

will not compile.

The compiler is not smart enough to ignore the unused lifetime at the moment.